neo/rawio/cedrawio.py

```diff
@@ -0,0 +1,177 @@
+"""
+Class for reading data from CED (Cambridge Electronic Design)
+http://ced.co.uk/
+
+This read *.smrx (and *.smr) from spike2 and signal software.
+
+Note Spike2RawIO/Spike2IO is the old implementation in neo.
+It still works without any dependency and should be faster.
+Spike2IO only works for smr (32 bit) and not for smrx (64 bit) files.
+
+This implementation depends on the SONPY package:
+https://pypi.org/project/sonpy/
+
+Please note that the SONPY package:
+  * is NOT open source
+  * internally uses a list instead of numpy.ndarray, potentially causing slow data reading
+  *  is maintained by CED
+
+
+Author : Samuel Garcia
+"""
+
+from .baserawio import (BaseRawIO, _signal_channel_dtype, _signal_stream_dtype,
+                _spike_channel_dtype, _event_channel_dtype)
+
+import numpy as np
+from copy import deepcopy
+
+try:
+    import sonpy
+    HAVE_SONPY = True
+except ImportError:
+    HAVE_SONPY = False
+
+
+class CedRawIO(BaseRawIO):
+    """
+    Class for reading data from CED (Cambridge Electronic Design) spike2.
+    This internally uses the sonpy package which is closed source.
+
+    This IO reads smr and smrx files
+    """
+    extensions = ['smr', 'smrx']
+    rawmode = 'one-file'
+
+    def __init__(self, filename='', take_ideal_sampling_rate=False, ):
+        BaseRawIO.__init__(self)
+        self.filename = filename
+
+        self.take_ideal_sampling_rate = take_ideal_sampling_rate
+
+    def _source_name(self):
+        return self.filename
+
+    def _parse_header(self):
+        assert HAVE_SONPY, 'sonpy must be installed'
+
+        self.smrx_file = sonpy.lib.SonFile(sName=str(self.filename), bReadOnly=True)
+        smrx = self.smrx_file
+
+        channel_infos = []
+        signal_channels = []
+        for chan_ind in range(smrx.MaxChannels()):
+            chan_type = smrx.ChannelType(chan_ind)
+            if chan_type == sonpy.lib.DataType.Adc:
+                physical_chan = smrx.PhysicalChannel(chan_ind)
+                divide = smrx.ChannelDivide(chan_ind)
+                if self.take_ideal_sampling_rate:
+                    sr = smrx.GetIdealRate(chan_ind)
+                else:
+                    sr = 1. / (smrx.GetTimeBase() * divide)
+                max_time = smrx.ChannelMaxTime(chan_ind)
+                first_time = smrx.FirstTime(chan_ind, 0, max_time)
+                # max_times is included so +1
+                time_size = (max_time - first_time) / divide + 1
+                channel_infos.append((first_time, max_time, divide, time_size, sr))
+                gain = smrx.GetChannelScale(chan_ind) / 6553.6
```

Comment:
that gain looks rather arbitrary. Are you sure this is consistent across files? Can it be retrieved from the header?

Reply:
It is in fact in the documentation and have been tested with the spike2 software with Eduarda.
The spike2io also have the same gain.